### PR TITLE
Adds a trending_towards attribute when period is set

### LINKED ISSUE #215: Add a trend attribute for sensors with a period
### Checklist

- [X] I have filled out the template to the best of my ability.
- [X] This only contains 1 feature request (if you have multiple feature requests, open one feature request for each feature request).
- [X] This issue is not a duplicate feature request of [previous feature requests](https://github.com/ludeeus/integration_blueprint/issues?q=is%3Aissue+label%3A%22Feature+Request%22+).

### Is your feature request related to a problem? Please describe.

Im using the average sensor to control my power usage as we get billed more if we exceed a certain average wattage per hour. 

### Describe the solution you'd like

To make the process explained above easier i would appreciate a "trending towards" attribute in the average sensor when a period is set that would function like:
 - If the current state of the monitored entities would remain the same for the remainder of the period my proposed "trending towards" attribute would report the expected state that the "average" sensor would have at the end of the period
 
Example:
If you are 1/4 of the way into the current period and the current state of the average sensor is 1000 and the current average of the entities is 2000
Then my proposed attribute would report 1000/4 + 2000 * 3/4 = 1750

This would enable me to make automations to reduce power draw or send me notifications if i am projected to exceed some limit i have configured. 
Lets say for example that i have configured a limit of 1500 for my example above. I would then be able to easily turn on/off heating etc to keep the "trending towards" attribute bellow that limit

### Describe alternatives you've considered

I could solve this in home assistant by using automations or a template sensor, but having the functionality directly in the sensor would be very neat.

### Additional context

no additional context

### PR DESCRIPTION
**Pull request by AnderzL7**
## Proposed change

This adds a new trending_towards attribute that has the value of the expected average at the end of the period if the monitored entities keeps their current states. This is calculated by the following script block where current_average is the current average of the monitored entities:

```py
current_average = round(
                sum(current_values) / len(current_values), self._precision
            )
if self._precision < 1:
    current_average = int(current_average)
part_of_period = (now_ts - start_ts) / (actual_end_ts - start_ts)
to_now = self._attr_native_value * part_of_period
to_end = current_average * (1 - part_of_period)
self.trending_towards = to_now + to_end
```

This new attribute can be used to predict what the resulting average of this period will be or in my case to programmatically limit my power usage below a limit i have set by using automations.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an this integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:

```yaml
# Example configuration.yaml
- platform: average
  name: Average power this hour
  entities:
    - sensor.emi_power
  start: "{{ now().replace(minute=0, second=0, microsecond=0) }}"
  end: "{{ now().replace(minute=59, second=59, microsecond=999) }}"
  unique_id: "average_power_h_hacs-2024-01"
```

## Additional information

- This PR fixes or closes issue: closes #215

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] The code has been formatted using Black (`black --fast custom_components`)

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated to README.md

### CODE REVIEW

File: custom_components/average/sensor.py [hunk 1]
```diff
@@ -535,6 +545,7 @@ async def _async_update_state(
                         last_elapsed = end_ts - last_time
                         value += last_state * last_elapsed
                         elapsed += last_elapsed
+                        current_values.append(last_state)
```
Comment (Limych) [inline custom_components/average/sensor.py]: This is potentially dangerous code.

In the next issue (#222), a person already complains that the component freezes the entire system when it tries to pass an array of 100,000 records through the component.

If I add your code, there is a high risk of total system slowdown at much smaller volumes.

Please try to solve the same problem without re-storing the entire data array in memory.

**Review by Limych — CHANGES_REQUESTED** [review]

File: custom_components/average/sensor.py (hunk 1 above)
Comment (AnderzL7) [reply custom_components/average/sensor.py]: Don't know why i did it this way :P Changed it to only store the last_state in its own variable and not all values.

Comment (AnderzL7) [reply custom_components/average/sensor.py]: Nevermind. I looked over my code again. It stores the last values for each entity in its own list to use the average of that list for the trend. I added a check to only append to the list if the last value is of instance number (just like your own code). I don't see how this new code is any more unsafe than your own. So if you still think it is unsafe, please explain how i can improve that

**Comment by Limych** [comment]
Thanks :heart:

**Pull request merged**